messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java

```diff
@@ -0,0 +1,192 @@
+    private final static String METHOD_PARAMS = "bix";
+    private final static String METHOD_RETURN = "bix";
+    private final Compressor compressor = new Compressor(CompressionType.LZ4, 3, 90, 1024);
+
+    @Override
+    protected String getReturnSpec() { return METHOD_RETURN; }
+    @Override
+    protected Method buildMethod() {
+
+        Method method = new Method(METHOD_NAME, METHOD_PARAMS, METHOD_RETURN, this);
+        method.methodDesc("Send a message bus request and get a reply back.");
+        method.paramDesc(0, "encoding", "Encoding type.")
+                .paramDesc(1, "decodedSize", "Number of bytes after decoding.")
+                .paramDesc(2, "payload", "Slime encoded payload.");
+        method.returnDesc(0, "encoding", "Encoding type.")
+                .returnDesc(1, "decodedSize", "Number of bytes after decoding.")
+                .returnDesc(2, "payload", "Slime encoded payload.");
+        return method;
+    }
+    private static final String VERSION_F = new String("version");
+    private static final String ROUTE_F = new String("route");
+    private static final String SESSION_F = new String("session");
+    private static final String PROTOCOL_F = new String("prot");
+    private static final String TRACELEVEL_F = new String("tracelevel");
+    private static final String TRACE_F = new String("trace");
+    private static final String USERETRY_F = new String("useretry");
+    private static final String RETRY_F = new String("retry");
+    private static final String RETRYDELAY_F = new String("retrydelay");
+    private static final String TIMEREMAINING_F = new String("timeleft");
+    private static final String ERRORS_F = new String("errors");
+    private static final String SERVICE_F = new String("service");
+    private static final String CODE_F = new String("code");
+    private static final String BLOB_F = new String("msg");
+    private static final String MSG_F = new String("msg");
+
+    @Override
+    protected Request encodeRequest(Version version, Route route, RPCServiceAddress address, Message msg,
+                                    long timeRemaining, byte[] payload, int traceLevel) {
+        Slime slime = new Slime();
+        Cursor root = slime.setObject();
+
+        root.setString(VERSION_F, version.toString());
+        root.setString(ROUTE_F, route.toString());
+        root.setString(SESSION_F, address.getSessionName());
+        root.setString(PROTOCOL_F, msg.getProtocol().toString());
+        root.setBool(USERETRY_F, msg.getRetryEnabled());
+        root.setLong(RETRY_F, msg.getRetry());
+        root.setLong(TIMEREMAINING_F, msg.getTimeRemaining());
+        root.setLong(TRACELEVEL_F, traceLevel);
+        root.setData(BLOB_F, payload);
+
+        byte[] serializedSlime = BinaryFormat.encode(slime);
+        Compressor.Compression compressionResult = compressor.compress(serializedSlime);
+        Request req = new Request(METHOD_NAME);
+        Values v = req.parameters();
```

Comment:
Assuming we want to move to TLS and to be able to extend request headers with secret-bearing Authorization headers etc in the future, compressing these together with an arbitrary payload is unfortunately not secure. See https://en.wikipedia.org/wiki/CRIME and https://en.wikipedia.org/wiki/BREACH.

tl;dr assuming an adversary can observe encrypted packets and cause the system to send packets containing own data (user feed/queries is a very realistic, non-theoretical case), the resulting compressed length of the ciphertext is used as an oracle that allows you to iteratively deduce  secret contents in _O(n)_ rather than _O(2^n)_.

Separating headers from payload and only compressing the latter will at least ensure we don't leak our own secrets.

Reply:
use body, and allow for later header payload.